Oblivious LRU with delete -- addressed pull request comments, added LRUMapWithDelete

mrflip opened the pull request:
This PR is all to the credit of @trivikr's PR #162 ; in order to help accelerate its inclusion in the proper release I've added commits that address the PR comments, and also added LRUMapWithDelete using the analogous strategy.

It wasn't clear to me what ObliviousLRUCache meant -- perhaps it's a term of art I haven't met yet -- but I took the liberty of renaming it to the homely but very clear LRUCacheWithDelete / LRUMapWithDelete. (Also, putting the adjective last grouped them together in the tests / indexes / etc where things are alphabetized). I'm glad to redo the commit history if ObliviousLRUCache is the proper name for this widget.

Last point -- in the comments on 162 @Yomguithereal recommended "a rolling test ... add 2 remove 1, in a loop of ~10 items for a cache of ~4-5 capacity." I tried to think of a way to give it a workout with a nice clean loop and still test it meaningfully -- like a fizzbuzz kind of thing -- but bringing in a pattern seemed too artificial and I think there'd be a lot of logic lying around. Instead, I just spammed in a whole bunch of inserts and deletes and asserted the state of the cache at various rest stops. To my mind this proved the same point that a loop would but more clearly demonstrated there was nothing up my sleeve mathematically. Does this do the trick for you?

mrflip commented:
One other change from 162 -- at the cost of one extra array reference op, the delete method returns the dead item. If that cost is meaningful I can add a ¿deletePop? method to parallel set / setPop.

Yomguithereal commented on lru-cache-with-delete.js:
```diff
@@ -0,0 +1,201 @@
+    return undefined;
+  }
+
+  const dead = this.V[pointer];
```
I don't have any issue with this particularly, and I don't think the array deref is such a perf hit but I notice that the equivalent in, say, an `ES6` map, is to return a boolean on delete. I agree it looks more useful to return the value even though you won't be able to distinguish between an actual deletion or else if you use `undefined` as values (but if you do so, who are you? why are you doing this :) ?).

mrflip replied:
undefined is undefined, anyone who wishes to distinguish should spell it n-u-l-l.

I'll update to add the "returns its dead" method as remove, along with a delete method that `!!`ings it into a boolean. If someone feels strongly about the extra *two* ops I've now added to delete they can unDRY it in a future pull request.

Yomguithereal commented:
Thanks for your PR @mrflip. I have opened a small review with some comments. The unit tests look good to me.

mrflip commented on lru-cache-with-delete.js:
```diff
@@ -0,0 +1,201 @@
+ * @return {undefined}
+ */
+LRUCacheWithDelete.prototype.set = function(key, value) {
+  this.setpop(key, value);
```
Even if I benched it fine on say Node, I wouldn't trust that to extend to whatever browsers. I propose to un-DRY it as you did for the parent class.

mrflip commented:
Pushed a new commit addressing your comments. Remove lets you supply a singleton -- eg `Symbol('missing')` -- to indicate absence, permitting one to have undefined be in range.

Yomguithereal commented:
Looks good to me @mrflip. I am trying to find a less conspicuous name for the structure but I don't suceed, so I guess explicit is nice. @trivikr any remark before I merge and release?

Yomguithereal commented:
I merged the PR just now. Will release a new version in the evening I think (I need to batch up some SparseSet improvements along with it).

Yomguithereal commented:
v0.39 is now live
